### PR TITLE
Feature/liquidstaking2.0

### LINKED ISSUE #432: Redesigned liquid stake claiming.
**Storage**
* Matchingpool: maintain the total stake/unstake token.
* StakingSettlementHistory: Record slash/reward for each era. FIXME(Alan WANG): We should ensure no era miss.
* UnstakeQueue: `Vec<(who, amount)>` the queue to pay off.
* ExchangeRate

**Hooks**
* fn on_idle
  - Pay off the unstake request in the queue till pallet's balance insufficient.
 
**Call**:
* fn stake(origin, amount)
   - mint xtoken, transfer token to pallet.
   - acc `total_stake_token`
* fn unstake(origin, amount)
   - burn xtoken, try to transfer token to user, otherwise push it into a queue.
   - acc `total_unstake_token`
* fn record_settlement_response
   - applay reward/slash, and sync exchange rate.
* fn settlement(unbonding_token)
   - it's invoked by stake_client, and should pass how many unbonding token in relaychain currently.
   - calculate and deposit `(bond_amount, rebond_amount, unbond_amount)` as an event. And transfer `bond_amount` to relaychain at the same time.
   - reset total stake/unstake token.

-------

Pros:
- Regardless stability of stake-client. It works even if stake-client broken for a while.

### PR DESCRIPTION
**Pull request by alannotnerd**
Close #432

### CODE REVIEW

File: pallets/liquid-staking-v2/Cargo.toml [hunk 1]
```diff
@@ -14,6 +14,8 @@ frame-system       = { git = 'https://github.com/paritytech/substrate.git', bran
 sp-runtime         = { git = 'https://github.com/paritytech/substrate.git', branch = 'polkadot-v0.9.8', version = '3.0.0', default-features = false }
 sp-std             = { git = 'https://github.com/paritytech/substrate.git', branch = 'polkadot-v0.9.8', version = '3.0.0', default-features = false }
 orml-traits        = { git = 'https://github.com/open-web3-stack/open-runtime-module-library.git', default-features = false }
+xcm                = { git = 'https://github.com/paritytech/polkadot.git', branch = 'release-v0.9.8', default-features = false }
+max-encoded-len    = { git = 'https://github.com/paritytech/substrate', branch = 'polkadot-v0.9.8', default-features = false }
```
Comment (GopherJ) [inline pallets/liquid-staking-v2/Cargo.toml:18]: let's use codec's `max-encoded-len` feature

Comment (alannotnerd) [reply pallets/liquid-staking-v2/Cargo.toml:18]: Fix it in next commit.

**Review by MrPai — COMMENTED** [review]
I need to think about this new design, besides, we could hear the other's suggestion.

File: pallets/liquid-staking-v2/src/lib.rs [hunk 2]
```diff
@@ -216,8 +200,51 @@ mod pallet {
         }
     }
 
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+        fn on_idle(_n: BlockNumberFor<T>, remaining_weight: Weight) -> Weight {
+            let mut remaining_weight = remaining_weight;
+            let base_weight = T::WeightInfo::pop_queue();
+            loop {
```
[the remark this answers is not in the document]
Comment (alannotnerd) [reply pallets/liquid-staking-v2/src/lib.rs:208]: Not sure there is a better solution to iter over a vector, pop front and break while specific condition satisfied.

[the remark this answers is not in the document]
Comment (alannotnerd) [reply pallets/liquid-staking-v2/src/lib.rs:208]: @kaichaosun

File: pallets/liquid-staking-v2/src/types.rs [hunk 3]
```diff
@@ -41,3 +19,26 @@ pub struct MatchingLedger<Balance> {
     /// **NOTE** will be calculated by: exchangeRate * xToken amount
     pub total_unstake_amount: Balance,
 }
+
+impl<Balance> MatchingLedger<Balance>
+where
+    Balance: AtLeast32BitUnsigned + Copy + Clone,
+{
```
Comment (yz89) [inline pallets/liquid-staking-v2/src/types.rs:26]: add docs for args/return value.

File: pallets/liquid-staking-v2/src/lib.rs [hunk 4]
```diff
@@ -347,29 +347,39 @@ mod pallet {
...
-                        status: ResponseStatus::Succeeded,
-                        ..op
-                    })
-                    .ok_or(Error::<T>::OperationNotReady)?;
-                *op = Some(next_op);
-                Ok(())
-            })?;
+            let (bond_amount, rebond_amount, unbond_amount) =
+                MatchingPool::<T>::take().matching(unbonding_amount);
+            T::XcmTransfer::transfer(
```
Comment (yz89) [inline pallets/liquid-staking-v2/src/lib.rs:365]: The xcm::transfer should only be called once you get a Bond request.

File: pallets/liquid-staking-v2/Cargo.toml (hunk 1 above)
Comment (GopherJ) [reply pallets/liquid-staking-v2/Cargo.toml:18]: you can click resolve if it's solved

**Comment by GopherJ** [comment]
let's switch to polkadot-v0.9.9

File: pallets/liquid-staking-v2/Cargo.toml [hunk 5]
```diff
@@ -8,12 +8,13 @@ version = '3.0.0'
...
 
 [dependencies]
-codec         = { package = 'parity-scale-codec', version = '2.0.0', default-features = false }
+codec              = { package = 'parity-scale-codec', version = '2.0.0', features= ["max-encoded-len"], default-features = false }
 frame-support = { git = 'https://github.com/paritytech/substrate.git', branch = 'polkadot-v0.9.9', default-features = false }
 frame-system  = { git = 'https://github.com/paritytech/substrate.git', branch = 'polkadot-v0.9.9', default-features = false }
 orml-traits   = { git = 'https://github.com/open-web3-stack/open-runtime-module-library.git', default-features = false }
 sp-runtime    = { git = 'https://github.com/paritytech/substrate.git', branch = 'polkadot-v0.9.9', default-features = false }
 sp-std        = { git = 'https://github.com/paritytech/substrate.git', branch = 'polkadot-v0.9.9', default-features = false }
+xcm                = { git = 'https://github.com/paritytech/polkadot.git', branch = 'release-v0.9.9', default-features = false }
```
Comment (GopherJ) [inline pallets/liquid-staking-v2/Cargo.toml:17]: let's format it properly like other Cargo.toml

File: pallets/liquid-staking-v2/Cargo.toml [hunk 6]
```diff
@@ -8,12 +8,13 @@ version = '3.0.0'
 targets = ['x86_64-unknown-linux-gnu']
 
 [dependencies]
-codec         = { package = 'parity-scale-codec', version = '2.0.0', default-features = false }
+codec              = { package = 'parity-scale-codec', version = '2.0.0', features= ["max-encoded-len"], default-features = false }
```
Comment (GopherJ) [inline pallets/liquid-staking-v2/Cargo.toml:11]: use  single quote

File: pallets/liquid-staking-v2/src/lib.rs [hunk 7]
```diff
@@ -216,8 +204,57 @@ mod pallet {
...
+                if Self::unstake_queue().is_empty() {
+                    break;
+                }
+
+                let (who, amount) = Self::unstake_queue()
+                    .first()
+                    .expect("Has been checked before.")
+                    .clone();
+
+                if T::Currency::transfer(
```
Comment (GopherJ) [inline pallets/liquid-staking-v2/src/lib.rs:233]: we should define a variable here instead of writing it all together

**Review by yz89 — APPROVED** [review]

File: pallets/liquid-staking-v2/src/lib.rs [hunk 8]
```diff
@@ -86,6 +87,15 @@ mod pallet {
...
         type PalletId: Get<PalletId>;
 
+        /// XCM transfer
+        type XcmTransfer: XcmTransfer<Self::AccountId, Balance, CurrencyId>;
+
+        /// Base xcm transaction weight
+        type BaseXcmWeight: Get<Weight>;
+
+        /// Account manages the staking assets.
+        type Agent: Get<Self::AccountId>;
```
Comment (GopherJ) [inline pallets/liquid-staking-v2/src/lib.rs:97]: we should make it clear that this agent should be on relaychain, also we should prefer `MultiLocation`

File: pallets/liquid-staking-v2/src/lib.rs [hunk 9]
```diff
@@ -216,8 +204,57 @@ mod pallet {
...
+                    &who,
+                    amount,
+                )
+                .is_err()
+                {
+                    break;
+                }
+
+                remaining_weight -= base_weight;
+                UnstakeQueue::<T>::mutate(|v| {
```
Comment (GopherJ) [inline pallets/liquid-staking-v2/src/lib.rs:245]: can we make this a function  to make it more meaningful

Comment (alannotnerd) [reply pallets/liquid-staking-v2/src/lib.rs:245]: I could add some more comments.

File: pallets/liquid-staking-v2/src/lib.rs (hunk 7 above)
Comment (alannotnerd) [reply pallets/liquid-staking-v2/src/lib.rs:234]: I think it's already simple and clear enough. I'll add more comment if you warried.

**Pull request merged**